src/firmware/main/math/polynomial_2d.c

```diff
@@ -0,0 +1,175 @@
+GENERATE_2D_POLYNOMIAL_GET_VALUE_FUNCTION_DEFINITION(2)
+GENERATE_2D_POLYNOMIAL_GET_VALUE_FUNCTION_DEFINITION(1)
+GENERATE_2D_POLYNOMIAL_GET_VALUE_FUNCTION_DEFINITION(0)
+
+#define GENERATE_2D_POLYNOMIAL_DIFFERENTIATE_FUNCTION_DEFINITION(FROM_ORDER, TO_ORDER)   \
+    Polynomial2dOrder##TO_ORDER##_t shared_polynomial2d_differentiateOrder##FROM_ORDER(  \
+        Polynomial2dOrder##FROM_ORDER##_t p)                                             \
+    {                                                                                    \
+        Polynomial2dOrder##TO_ORDER##_t derivative = {                                   \
+            .x = shared_polynomial1d_differentiateOrder##FROM_ORDER(p.x),                \
+            .y = shared_polynomial1d_differentiateOrder##FROM_ORDER(p.y),                \
+        };                                                                               \
+        return derivative;                                                               \
+    }
+GENERATE_2D_POLYNOMIAL_DIFFERENTIATE_FUNCTION_DEFINITION(3, 2)
+GENERATE_2D_POLYNOMIAL_DIFFERENTIATE_FUNCTION_DEFINITION(2, 1)
+GENERATE_2D_POLYNOMIAL_DIFFERENTIATE_FUNCTION_DEFINITION(1, 0)
+
+
+#define GENERATE_2D_POLYNOMIAL_GET_ARC_LENGTH_PARAMETRIZATION_FUNCTION_DEFINITION(         \
+    N, N_minus_1)                                                                          \
+    void shared_polynomial_getArcLengthParametrizationOrder##N(                            \
+        Polynomial2dOrder##N##_t p, float t_min, float t_max,                              \
+        ArcLengthParametrization_t parametrization)                                        \
+    {                                                                                      \
+        assert(t_min < t_max);                                                             \
+                                                                                           \
+        if (parametrization.num_values == 1)                                               \
+        {                                                                                  \
+            parametrization.s_values[0] = 0;                                               \
+            parametrization.t_values[0] = (t_max + t_min) / 2.0;                           \
+            return;                                                                        \
+        }                                                                                  \
+                                                                                           \
+        Polynomial2dOrder##N_minus_1##_t deriv =                                           \
+            shared_polynomial2d_differentiateOrder##N(p);                                  \
+                                                                                           \
+        float dt = (t_max - t_min) / ((parametrization.num_values - 1) * 2);               \
```

Comment:
Just to clarify, is `dt` being divided by 2 to ensure that the number of `dt` is even for Simpson's rule? (And increase in resolution for integration approximation?)

Reply:
Yeah, that's exactly correct. I'll add a comment here indicating as such.